Update routing hints upon insufficient fees failure.

Bluetegu opened the pull request:
This pull request address issue #2066 

Reconstruct the problem:
Setup composed of 3 nodes, A -> B -> C, where B -> C is private.
Create an invoice on C   (C:addinvoice with private flag)
Increase fee on B->C   (B:updatechanpolicy to a higher fee)
Attempt to pay invoice from A  (A:sendpayment --pay_req=xxx)
Receive payment error ("unable to route payment to destination: FeeInsufficient ... ")

After the fix the payment goes through.

A unit test on the 'basic graph' was added. The test adds a private channel between son goku and elst, with lower fees compared with the alternative routes from roasbeef to elst. A fee error is returned on first attempt to route through the private channel, with a higher fee, but still better compared with the alternative routes. The test checks that the after the update the payment is still routed through the private channel but with the updated higher fees.

A bug in the existing test TestSendPaymentErrorRepeatedFeeInsufficient was fixed too. I did a minimal fix. I believe this test can be further improved. The major issue was that the simulated error returned was rejected due to signature failure, and hence didn't simulate correctly the insufficient fees error as intended. In order to correctly sign the errors, an optional private key attribute was added to the basic graph and its parser.

joostjager commented on routing/router.go:
```diff
@@ -1838,6 +1840,8 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 			// newly updated fees.
 			case *lnwire.FailFeeInsufficient:
```
This is not the only case where we can get back a channel update. For example, `FailIncorrectCltvExpiry`. I think we should try to update `additionalEdges` also in those cases.

joostjager replied:
Can the update maybe be done inside `applyChannelUpdate` ?

Bluetegu replied:
Done.

joostjager commented on routing/router_test.go:
```diff
@@ -521,7 +548,7 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
 
 	// Craft a LightningPayment struct that'll send a payment from roasbeef
-	// to luo ji for 100 satoshis.
+	// to sophon for 1000 satoshis.
```
I think the fix to this existing test can be moved to a separate commit, stacking on top of the priv key map commit.

Bluetegu replied:
I prefer to leave it as is. Its a pain to extract those changes, and I think the fix is aligned with the signature checks that were added to the new test, so its not an unrelated. Anyhow, if you believe it's important let me know and I'll extract to a separate commit.

joostjager replied:
If you remove the mission control policy update code and `TestSendPaymentErrorFeeInsufficientPrivateEdge` (and leaving `signErrChanUpdate` in), I think a clean commit remains that fixes the existing test. And then commit on top of that the new functionality and its tests.

In know spending time on clean commit history can feel like a pain. But a large part of it is getting used to it, using the right tools and understanding git rebase commands. I use vscode for this sometimes. With all changes in the working directory, it allows you to select chunks of files and stage only those. There are other editors that offer similar functionality.

The basic assumption here is that it indeed takes the author of the PR extra time (but with some experience this can be marginal) than just creating a single big commit, but more time is saved by the (at least two) people who review the code and all future developers looking up this commit to understand the greater picture of a particular change.

Bluetegu replied:
Ok. Done

joostjager replied:
Thanks. For the non-author it really makes it much easier to understand what parts of the full set of changes relate to each other.

joostjager commented on routing/pathfind_test.go:
```diff
@@ -74,11 +75,14 @@ type testGraph struct {
-	Alias  string `json:"alias"`
+	Source  bool   `json:"source"`
+	PubKey  string `json:"pubkey"`
+	PrivKey string `json:"privkey"`
```
Addition of a private key to the json can be a separate commit.

Bluetegu replied:
Done

joostjager commented on routing/router_test.go:
```diff
@@ -602,13 +639,157 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	ctx.router.cfg.SendToSwitch = func(firstHop lnwire.ShortChannelID,
+		_ *lnwire.UpdateAddHTLC, _ *sphinx.Circuit) ([32]byte, error) {
+
+		if !errorReturned {
```
Flipping the boolean looks quite implementation-dependent. Is it possible to do it in a more robust way? For example by inspecting the contents of the `UpdateAddHTLC` and matching with the updated policy?

Bluetegu replied:
What I wanted there is a simple way to send an error only the first time; that is, if its wasn't Go, I would have a static variable named say 'once' and change it to True inside the function. I think inspecting the content of UpdateAddHTLC would obfuscates this. Maybe the name errorReturned is confusing? Is there a better way in Go to achieve this 'static' variable 'do once' logic?

joostjager replied:
Oh no, I don't mean that there is anything wrong with this code on the language level. 

This SendToSwitch is supposed to simulate the real behaviour to cover the private channel policy update case. If you flip the boolean, you rely on the fact the two htlcs will be sent out in a particular order. That's what I meant by '(router) implementation specific'. I was just thinking, if you just check the total amount in `UpdateAddHtlc` and decide based on that to reject or not, it is a little more robust.

Bluetegu replied:
I don't agree here.
This is a unit test that orchestrate a scenario in which the first attempt to pay returns an error while the second one does not.
It's explicitly designed to fail only once, and not say, to fail twice and then succeed. 
Once the second attempt succeeds, the resulting route is checked, including a check that the same route is taken, and that the fee was indeed updated according to the error simulated.
I do 'rely on the fact the two htlcs will be sent out in a particular order'. Anything else will be detected as error. 
If instead I base the check on values in UpdateAddHtlc, I obfuscate the scenario I wanted to simulate (error only once), I have to avoid loops if the expected values do not appear, have to count the number of attempts to send to switch and compare later to 2, etc.

joostjager replied:
It indeed depends on what exactly you want to assert. Ok, no change needed then.

joostjager commented on routing/router_test.go:
```diff
@@ -602,13 +639,157 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	t.Parallel()
+
+	const startingBlockHeight = 101
+	ctx, cleanUp, err := createTestCtxFromFile(startingBlockHeight, basicGraphFilePath)
```
There are currently two ways of writing these kind of path finding tests. One is the way you did it, using the json file. The other option is used in for example `TestChannelUpdateValidation`. It allows programmatically building the test graph and generates some values like the keys by itself (as the actual value isn't important for the tests, it just needs a value). You could consider using that instead. I think it is clearer than keep adding new channels to the json file.

Bluetegu replied:
I think that you will still have the problem of needing a private key even if you use TestChannelUpdateValidation approach, as the validation of the error message is part of the tested code. Anyhow, I like the approach of using the basic graph in many tests all based on a non trivial topology. It requires the writer of the test to rethink whether the logic indeed covers all scenarios, beyond the simplest topology needed for testing the particular feature or fix she was aiming to do.

joostjager replied:
That's okay, just wanted to mention the option.

Bluetegu commented:
Thanks for the review!
I rebased and changed the commit message(s). The code in router.go looks much better all around.
I addressed most of the comments. Please see responses to individual comments inline.

Bluetegu commented on routing/missioncontrol.go:
```diff
@@ -291,6 +291,27 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
 
+// UpdateEdgePolicy updates the edge policy of the routing hints kept in the
+// payment session. The number of hints updated is returned.
+func (p *paymentSession) UpdateEdgePolicy(e *channeldb.ChannelEdgePolicy) uint32 {
```
Done

Bluetegu commented:
I rebased again, but for some reason the third CI test (coverage) fails in auto pilot and I don't think this has anything to do with the code modification in this PR. I forced push twice already and got the same error there.

joostjager commented on routing/router.go:
```diff
@@ -1818,6 +1820,12 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 
+				err := ValidateChannelUpdateAnn(pubKey, update)
+				if err != nil {
+					log.Errorf("Unable to validate channel update: %v", err)
```
This line and others too exceed the max line width of 80 chars (project coding standard).

Bluetegu replied:
The log line is of 77 width. I believe I kept the <80 width everywhere.

joostjager replied:
I have it at 96 chars. Do you have tab size 8?

joostjager replied:
The standard is 80 chars max @ tab size 8.

joostjager commented on routing/router.go:
```diff
@@ -1818,6 +1820,12 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 				update *lnwire.ChannelUpdate,
 				pubKey *btcec.PublicKey) {
 
+				err := ValidateChannelUpdateAnn(pubKey, update)
```
Why is this added here? Validation is also done inside `r.applyChannelUpdate`.

Bluetegu replied:
I agree this is ugly. My excuse it that I'm trying to minimize the changes I do in code to those mandatory for this PR. Therefore, if validation fails, processChannelUpdateAndRetry should immediately return, and should not update paySession at all (neither update hints or report edge policy failure), so the check must be there. However we can not rely on the error returned by applyChannelUpdate, as it may well error in other scenarios (i.e. not validation errors) which are perfectly ok, and do require retries. Initially I thought of doing the validation much earlier, i.e. before the switch on the error types, but this looks like a change that is too major for this PR; hence my feeling was that adding validation here is the right way to go, with only bad consequence being that its checked twice.

joostjager replied:
I think it is ok to not update the payment session if the graph update fails. `ReportEdgeFailure` is important to call in case of a validation error. If we get an incorrect update we don't want to try that channel anymore.

Bluetegu replied:
Ok. I removed the extra validation and returned if applyChannelUpdate fails after the edge is pruned.

I'm uncomfortable with the handling of error messages that fail in signature validation, as they do prune edges. Doesn't that allow an attacker to send error messages and cause the routing to fail?

joostjager commented on routing/router.go:
```diff
@@ -1834,6 +1842,21 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
+						FeeBaseMSat:               uBaseFee,
+						FeeProportionalMillionths: uFeeRate,
+					})
+
```
You've added it in here, but for other cases where `r.applyChannelUpdate` is called directly, the payment session isn't updated.

Bluetegu replied:
My idea is that the hints are updated only when a retry is aimed at. It now handles all those cases (not only insufficient fees). Otherwise the vertex or channel is removed completely, and there is no need for the update. I thought that is the whole purpose of defining processChannelUpdateAndRetry (which includes the paySession in its closure). 
Said in another way, it doesn't seem right to me to plug it into applyChannelUpdate, exactly because this function is called for changing database edges/vertexes (and don't have paySession in closure).

joostjager replied:
Yes, I think you're right. All other calls to `applyChannelUpdate` prune immediately, so those edges will never show up in subsequent payment attempts anymore.

joostjager commented on routing/router_test.go:
```diff
@@ -110,6 +112,7 @@ func createTestCtxFromGraphInstance(startingHeight uint32, graphInstance *testGr
 		router:    router,
 		graph:     graphInstance.graph,
 		aliases:   graphInstance.aliasMap,
+		privKeys:  graphInstance.privKeyMap,
```
This looks like it belongs more to the previous commit where priv keys are introduced.

Bluetegu replied:
Fixed

Bluetegu replied:
Done

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +299,22 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+	// run over hints and update policy if edge id matches
+	for _, edges := range p.additionalEdges {
+		for _, edge := range edges {
+			if edge.ChannelID == e.ChannelID {
```
Are you updating the channel in both directions? I think you may also need to match the pubkey in additional edges to update only a single direction.

Bluetegu replied:
The channel is updated only in one direction. That is, the node generates an error on the outgoing channel policy, and you update the hints accordingly. Am I missing something here?

joostjager replied:
I mean, suppose you have two additional edges relating to the same channel id. They will both be updated?

Bluetegu replied:
Yes. I changed it such that it takes direction into account too, such that policy will be updated only if channel ID and direction both match.
I think this also fixes a general hidden bug in routing through hint hops, not related to this PR, as additionalEdges policies didn't take into account direction. That is, the Flags were not initialized in NewPaymentSession, and routing assumes the direction is initialized in all edge policies.

Bluetegu commented:
Ready for inspection :-)
By some miracle the cover test passed too.
Please see inline my response to the review issues.
Thanks!

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +306,25 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+			if edge.ChannelID == e.ChannelID &&
+				edge.Flags&lnwire.ChanUpdateDirection ==
+					e.Flags&lnwire.ChanUpdateDirection {
+				edge.FeeBaseMSat = e.FeeBaseMSat
```
New line after wrapped line

Bluetegu replied:
Done

joostjager commented on routing/router.go:
```diff
@@ -1832,7 +1834,30 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
+				var direction lnwire.ChanUpdateFlag
+				if failedEdge.direction == 1 {
+					direction = lnwire.ChanUpdateDirection
+				}
```
`failedEdge.direction` can be used directly.

Bluetegu replied:
Not sure. failedEdge direction is unit8 while Flags are unit16.

joostjager replied:
Ok, you may need to do a type cast, but the if statement is not necessary.

Bluetegu replied:
Done

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +306,25 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+// UpdateEdgePolicy updates the edge policy of the routing hints kept in the
+// payment session.
+func (p *paymentSession) UpdateEdgePolicy(e *channeldb.ChannelEdgePolicy) {
+
```
remove newline

Bluetegu replied:
Done

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +306,25 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+// payment session.
+func (p *paymentSession) UpdateEdgePolicy(e *channeldb.ChannelEdgePolicy) {
+
+	// run over hints and update policy if edge id and direction matches
```
Capital at beginning of sentence. Period at end.

Bluetegu replied:
Fixed

joostjager commented on routing/missioncontrol.go:
```diff
@@ -209,6 +210,12 @@ func (m *missionControl) NewPaymentSession(routeHints [][]HopHint,
+			v := NewVertex(hopHint.NodeID)
+			var direction lnwire.ChanUpdateFlag
+			if bytes.Compare(v[:],
+				endNode.PubKeyBytes[:]) == 1 {
```
Can fit on a single line

Bluetegu replied:
Fixed

joostjager commented on routing/router.go:
```diff
@@ -1203,14 +1203,16 @@ func (r *ChannelRouter) processUpdate(msg interface{}) error {
-		}
+		if exists {
+			if err = r.cfg.Graph.UpdateEdgePolicy(msg); err != nil {
+				err := errors.Errorf("unable to update channel policy: %v", err)
```
Exceeds line len

Bluetegu replied:
Fixed

joostjager commented on routing/router.go:
```diff
@@ -1203,14 +1203,16 @@ func (r *ChannelRouter) processUpdate(msg interface{}) error {
-			log.Error(err)
-			return err
-		}
+		if exists {
```
Is this a fix in itself? Ideally this could be yet another separate commit, but if you find it too small for that, it's fine with me.

halseth replied:
I _think_ maybe the proper fix here is something like this: https://github.com/lightningnetwork/lnd/pull/2279/commits/ad92e5c9638c95b53f8756d462424d6317de76bf (feel free to cherry-pick it into this PR if needed).

I believe then you wouldn't have to flip the `AssumeChannelValid` in the config during tests?

halseth replied:
This can be removed now that #2549 is merged.

joostjager commented on routing/router.go:
```diff
@@ -1832,7 +1834,30 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
+				uBaseFee := lnwire.MilliSatoshi(update.BaseFee)
+				uFeeRate := lnwire.MilliSatoshi(update.FeeRate)
+				uMinHTLC := lnwire.MilliSatoshi(
+					update.HtlcMinimumMsat)
```
This vars are introduced because of line length I assume. But maybe that indicates that now is the time to extract this closure to a separate function.

joostjager replied:
There are plenty of internal functions that are called only once, it does help to make the code more understandable. Inline functions can be useful especially if a lot of local vars are bound.

How about just creating a function taking `failedEdge` and `update` and returning the `ChannelEdgePolicy`? Or better, updating the signature of `UpdateEdgePolicy` to take  `failedEdge` and `update`. Also consistent with other pay session funcs that take an `edgeLocator` too.

joostjager commented on routing/router_test.go:
```diff
@@ -647,6 +647,150 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+
+	// The route should pass via the private channel
+	if !route.containsChannel(privateChannelID) {
+		t.Fatalf("route did not pass through private channel between pham nuwen " +
```
Line length.

Bluetegu replied:
Fixed

joostjager commented on routing/missioncontrol.go:
```diff
@@ -219,9 +226,9 @@ func (m *missionControl) NewPaymentSession(routeHints [][]HopHint,
 					hopHint.FeeProportionalMillionths,
 				),
 				TimeLockDelta: hopHint.CLTVExpiryDelta,
+				Flags:         direction,
```
This was indeed not set, but I think it didn't have impact before, because that flag was never read right?

Bluetegu replied:
I think it did have impact, i.e. it was an open bug, but the situations where it matters were rare enough such that it wasn't noticed. I haven't fully thought this out. What happened was that when I tried to fix UpdateEdgePolicy to consider direction as well, following your comment, the (new) test failed, and adding direction here fixed it.

joostjager replied:
Yes, it was an omission that surfaced when you fixed `UpdateEdgePolicy`. But I think that before that, the situations where it mattered where non-existent. Just interested to know whether that was true. In any case, it is fixed now.

Bluetegu commented on routing/router_test.go:
```diff
@@ -162,6 +165,30 @@ func createTestCtxFromFile(startingHeight uint32, testGraph string) (*testCtx, f
 
+// Add valid signature to channel update simulated as error
+// received from the network.
+func signErrChanUpdate(key *btcec.PrivateKey,
```
Done.

Bluetegu commented:
Changes made.
The CI cover test failed, but I believe its a bug in the CI test itself (there is an open PR that fixes the coverage test).
Please see inline my response to the review issues.
Thanks!

Bluetegu commented:
Getting closer...
Open issues addressed. Please see inline.
Thanks!

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +305,33 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+	// Run over hints and update policy if edge id and direction match.
+	for _, edges := range p.additionalEdges {
+		for _, edge := range edges {
+			if edge.ChannelID == e.channelID &&
```
I think it could be slightly better to extract the if+update block into a separate function. Saves some indents and temp. vars.

Bluetegu replied:
Done

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +305,33 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+		for _, edge := range edges {
+			if edge.ChannelID == e.channelID &&
+				uint8(edge.Flags&lnwire.ChanUpdateDirection) ==
+					e.direction {
```
Newline after wrapped line.

Bluetegu replied:
Y

joostjager commented on routing/missioncontrol.go:
```diff
@@ -299,6 +305,33 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+// payment session.
+func (p *paymentSession) UpdateEdgePolicy(e *edgeLocator,
+	update *lnwire.ChannelUpdate) {
+	log.Debugf("Updating edge %v policy in Mission Control", e)
```
New line after wrapped line.

Bluetegu replied:
Y

joostjager commented on routing/router_test.go:
```diff
@@ -647,6 +647,150 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+				ctx.aliases))
+	}
+
+	// The route should pass via the private channel
```
Dot at end of sentence.

Bluetegu replied:
Y

joostjager commented on routing/router_test.go:
```diff
@@ -647,6 +647,150 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+			"between pham nuwen and elst")
+	}
+
+	// The route should have the updated fee
```
Dot at end of sentence.

Bluetegu replied:
Y

joostjager commented on routing/router_test.go:
```diff
@@ -647,6 +647,150 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+			"channel, instead got %v MSat", expectedFee,
+			route.HopFee(0))
+	}
+
```
Remove newline

Bluetegu replied:
Removed

joostjager commented:
Yes, almost there. Hope the quality level we try to maintain didn't discourage you too much!

Bluetegu commented:
Not at all. I have full appreciation for the need to maintain and improve quality. My experience till now was that the review process led to better end results, not only in term of code quality and conformance, but also in term of what the PR eventually achieves. For example this PR was written to address a bug of handling of insufficient fees error, and after review addresses all similar errors (all errors which lead to a second reroute attempt), as well as probably fixing another unrelated bug.

Changes in code made per review comments.
Thanks!

halseth commented on routing/missioncontrol.go:
```diff
@@ -299,6 +305,36 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+	ep *channeldb.ChannelEdgePolicy) {
+
+	if ep.ChannelID == el.channelID &&
+		uint8(ep.Flags&lnwire.ChanUpdateDirection) == el.direction {
```
This is a bit brittle as it assumes that the direction flag is the last bit (this is also true elsewhere in the codebase, but better not to keep assuming it) :)

Instead you can do `(ep.Flags&lnwire.ChanUpdateDirection != 0) == (el.direction != 0)` or similar.

Bluetegu replied:
Its strange that go doesn't have casting to bool in a simple way. Anyhow, I extracted it into the most straightforward help function, as to me this casting to bool is not self explanatory.

halseth commented on routing/missioncontrol.go:
```diff
@@ -219,9 +225,9 @@ func (m *missionControl) NewPaymentSession(routeHints [][]HopHint,
 					hopHint.FeeProportionalMillionths,
 				),
 				TimeLockDelta: hopHint.CLTVExpiryDelta,
+				Flags:         direction,
```
Ideally this would be fixed in its own commit to make it clear what the consequence of this change is.

Bluetegu replied:
The new test fails if the direction is not initialized here. I did try to add a unit test that directly checks this problem, but it didn't work as I expected; and I got into an impasse. 

I wanted the test to include two hints, one using a private channel in one direction and a second hint using the other direction of the same private channel. I wanted to have error on one of the directions, and check that the payment is routered through the other hint, through the private channel in a direction that didn't error. 

The setup was the following:  a graph including three nodes, a, b and c, connected by public edges in the following way.

               a  ------ b
               |
               |
               c

The payment should be routed to a node d. Node d is connected to b and c in private channels, and also c and b are connected in a private channel.

I included two hints   
h1: c -> b -> d
h2: b -> c -> d

I expected the route to be either 
r1 a -> c -> b -> d
r2 a -> b -> c -> d

But it seems that I'm getting either:
a1 a -> b -> d
a2 a -> c -> d

That is, I can't force routing to go through b -> c and c -> b, although that is what I explicitly indicate in the hints.
I was a bit surprised by that. Is routing expected to behave this way?

Initially I thought that I'd add negative fee on the b -- c channel to force routing to go through them, but unfortunately it doesn't seem that LND support negative fees at this point, as the fees are uint. 

I didn't have time to go deeper and make sure that I understand what is happening there correctly. I'll give it another try later.

Do you see a better test scenario?

Bluetegu replied:
I documented my finding regarding the above attempted test here #2346. 

My conclusion is that up till now in all legit scenarios there was no need for adding direction to additionalEdges during initialization in NewPaymentSession. This was only really required when direction was considered in UpdateEdgePolicy (both in missioncontrol). 
I don't recommend removing the direction initialization and checking, it is the right thing to do, but it doesn't solve any real life scenario, i.e. all real life scenarios were ok if both were omitted. 

IMO this means that there was no 'subtle bug' found (or its so subtle its invisible) and
that initializing the direction in NewPaymentSession doesn't deserve its own commit and its own unit test. It is part of the other logic that needs to address the original bug this PR is supposed to fix. The unit test that was added test the directionality already (it fails without initialization of the direction in NewPaymentSession) covers it.

I can add a test that attempt to send a malformed onion error that reports the wrong direction of the reported edge and check that it doesn't update the additionalEdges, but it seems too contrived. 

Let me know what you think.
Thanks!

halseth replied:
Thanks for the detailed explanation! 

So if I understand it correctly, the reason this hasn't been a problem is because we don't this flag when doing path finding? We infer the direction of the edge by looking at the to/from nodes I think.

One way to possibly test this is to look at the `ignoredEdges` map. This one uses `edgeLocators` which also takes into account directionality. So if you could add the first direction to this map, then include a routing hint for the second direction, you might be able to make it fail finding a path.

Bluetegu replied:
> So if I understand it correctly, the reason this hasn't been a problem is because we don't this flag when doing path finding? We infer the direction of the edge by looking at the to/from nodes I think.

Not exactly. What I'm saying is that if you didn't consider direction when pruning it'll be the same in real life scenarios.

Going back to the example detailed in the issue, suppose that we asked the router for 4 routes. 

We would get:
r1 A -> B -> D -> F
r2 A ->C -> E -> F
r3 A ->B -> D -> E -> F
r4 A ->C ->E -> D ->F

So it seems that we have both D->E and E->D on our list of candidates, and therefore we should make sure in code that if we receive an error on the D->E edge, we must prune only that direction D->E, but not E->D, hence the need for a direction check.

We should get an error on either directions of the E-D edge only once we attempt to route through it. We don't go through E-D in r1 and r2, so those need to fail first. 

So assume for our example that r1 failed due to D->F, so you prune it. and assume that r2 failed due to A->C, so you prune it too.

Now r3 is legit, so we try it. We pass through the D->E and we get an error.

With direction checking we prune D->E, but leave E->D. 

Without direction checking (as was before), we prune both E->D and D->E.

But r4 is useless anyhow (A->C already pruned), so it doesn't matter!

I'm quite confident that this can be formally proved for any graph whose cost at each hop is non-negative.

So bottom line, I don't see any real situation that I can simulate in a test that will require to check direction too, in addition to edge id, prior to pruning. I can create a contrived scenario, but as I said, the new test already test the direction checking anyhow.

halseth replied:
OK, I _think_ I see the problem you are facing testing this now. Here's a possible solution (using your original A-B-C-D graph):
* Include routing hints B-C-D and C-B-D
* Let the router try route A-B-C-D by making sure it is the cheapest path.
* Make the hop B-C fail, so it should be pruned.
* Now the router will pick either A-B-D or A-C-D, since it is impossible for A-C-B-D to be cheaper than both. Let's say it picks A-C-D. Make C-D fail.
* Now it picks A-B-D. Make A-B fail.
* The only possible route left is A-C-B-D. Before your bugfix this route wouldn't be found.

Bluetegu replied:
> Let the router try route A-B-C-D by making sure it is the cheapest path.
That was my original plan.
But, I can't. The router will select A->C->D
That is why I opened the issue #2346

halseth replied:
Well, what if you make the fees on channels A-C and B-D super high? If it is still choosing the routes A-C-D or A-B-D then, you found a bug.

Bluetegu replied:
Yes. I somehow convinced myself that its not possible. Anyhow, I separated this to a separate commit and added a dedicated unit test for that. Thanks!

halseth commented on routing/router_test.go:
```diff
@@ -647,6 +647,149 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+			"between pham nuwen and elst")
+	}
+
+	// The route should have the updated fee.
```
could also check the value of `errorReturned` here.

Bluetegu replied:
I didn't understand this one.

halseth replied:
You can check that `errorReturned == true`.

Bluetegu replied:
Ah. Added

halseth commented on routing/missioncontrol.go:
```diff
@@ -299,6 +305,36 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+	// Run over hints and update policy if edge id and direction match.
+	for _, edges := range p.additionalEdges {
+		for _, edge := range edges {
+			updateEdgePolicy(e, update, edge)
```
This is a bit hard to follow IMO. A better approach could be 
```go
for _, edges := range p.additionalEdges {
	for _, edge := range edges {
                if edge.ChannelID != e.ChannelID {
                      continue
                }

                if ..!checkDirection... {
                      continue
                }
 
                 ... update policy ....
        }
}
```
Also I don't think you need to pass the `edgeLocator` into this method? Might have everything you need in the `ChannelUpdate`. If not just pass the `pubKey` as in `applyChannelUpdate`.

Bluetegu replied:
Done

halseth commented on routing/router_test.go:
```diff
@@ -647,6 +647,149 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+// a fee related error from a private channel that we're attempting to route
+// through, then we'll update the fees in the route hints and successfuly
+// route through the private channel in the second attempt.
+func TestSendPaymentErrorFeeInsufficientPrivateEdge(t *testing.T) {
```
Good test! 👍

Bluetegu replied:
Thanks

Bluetegu commented on routing/missioncontrol.go:
```diff
@@ -209,6 +210,11 @@ func (m *missionControl) NewPaymentSession(routeHints [][]HopHint,
 			// and add it to list of edges corresponding to the node
 			// at the start of the channel.
+			v := NewVertex(hopHint.NodeID)
+			var direction lnwire.ChanUpdateFlag
```
Done

Bluetegu commented on routing/router_test.go:
```diff
@@ -566,7 +595,7 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
 		_ *lnwire.UpdateAddHTLC, _ *sphinx.Circuit) ([32]byte, error) {
 
-		roasbeefSongoku := lnwire.NewShortChanIDFromInt(chanID)
+		roasbeefSongoku := lnwire.NewShortChanIDFromInt(12345)
```
This is another (copy and paste) error in the test. Note that this is supposed to be roasbeefSongoku channel, not chanID that is channel edge from son goku to sophon.

halseth replied:
I see. The reason this test was passing before was that a signature error was returned instead?

Bluetegu replied:
Yes.

Bluetegu commented on routing/router_test.go:
```diff
@@ -583,7 +612,12 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
 	}
 
-	// Send off the payment request to the router, route through satoshi
+	// Avoid dropping the channel information upon check if the
```
I rephrased it. Hope this is clearer now.

Bluetegu commented on routing/pathfind_test.go:
```diff
@@ -190,6 +194,23 @@ func parseTestGraph(path string) (*testGraphInstance, error) {
+			return nil, err
+		}
+		if len(privBytes) > 0 {
+			key, derivedPub := btcec.PrivKeyFromBytes(btcec.S256(), privBytes)
```
Fixed

Bluetegu commented:
Thanks for the review!

Most comments addressed. I did try to create a new unit test to directly test the subtle bug found in the paySession initialization, but got into an impasse.

Please see inline.

halseth commented on routing/missioncontrol.go:
```diff
@@ -299,6 +305,46 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+			if edge.ChannelID != el.channelID {
+				continue
+			}
+			if edgePolicyDirection(edge) != el.direction {
```
We could also add a method for creating `edgeLocator` from a `ChannelUpdate`.

Bluetegu replied:
Added

halseth commented:
Blocked by #2279

Bluetegu commented:
@halseth  Hi Johan, Is the block tag necessary? Let's try to push this to completion. Thanks!

btcdanny commented:
> @halseth Hi Johan, Is the block tag necessary? Let's try to push this to completion. Thanks!

This would be great if possible, as we at Fastbitcoins are encountering the issue this PR addresses.

Bluetegu commented:
Ok, great. Let me know if anything is needed on my side.

halseth commented:
@Bluetegu #2549 is now merged, we can move this forward :)

Bluetegu commented:
@halseth Thanks!

halseth commented on routing/router_test.go:
```diff
@@ -587,7 +616,13 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	// before further processing an edge update. Toggel AssumeChannelValid
+	// to avoid dropping the channel update, as the utxo set query is
+	// not mocked in this test.
+	ctx.router.cfg.AssumeChannelValid = true
```
Is this still needed?

halseth commented on routing/router_test.go:
```diff
@@ -652,6 +653,373 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+// r3: C->E is pruned
+// leaving r4 as viable 4th route
+//
+// This test was added to test direction initialization in NewPaymentSession;
```
Excellent test description! 👍

halseth commented on routing/router_test.go:
```diff
@@ -652,6 +653,373 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	ctx, cleanUp, err := createTestCtxFromGraphInstance(startingBlockHeight,
+		testGraph)
+
+	defer cleanUp()
```
move `defer cleanUp()` after error check

halseth commented on routing/router_test.go:
```diff
@@ -652,6 +653,373 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	// before further processing an edge update. Toggel AssumeChannelValid
+	// to avoid dropping the channel update, as the utxo set query is
+	// not mocked in this test.
+	ctx.router.cfg.AssumeChannelValid = true
```
not needed anymore I think

halseth commented on routing/router_test.go:
```diff
@@ -652,6 +653,373 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	// not mocked in this test.
+	ctx.router.cfg.AssumeChannelValid = true
+
+	// Send payment
```
nit: missing period.

halseth commented on routing/router_test.go:
```diff
@@ -652,6 +653,373 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	}
+
+	// Now switch directions and test again.
+	ctx.router.missionControl.ResetHistory()
```
Instead of manually driving the test twice, it should be made table driven, such that the two scenarios can be run by the same logic.

That being said, this is an excellent test! 😀

halseth commented on routing/missioncontrol.go:
```diff
@@ -184,7 +183,7 @@ func (m *missionControl) NewPaymentSession(routeHints [][]HopHint,
 			// at the start of the channel.
 			v := NewVertex(hopHint.NodeID)
-			var flags lnwire.ChanUpdateFlag
+			var flags lnwire.ChanUpdateChanFlags
```
Needs a rebase I guess?

halseth commented on routing/payment_session.go:
```diff
@@ -35,6 +35,51 @@ type paymentSession struct {
 
+// edgeLocatorOfEdgePolicy returns the edge locator corresponding to
+// the channel edge policy
+func edgeLocatorOfEdgePolicy(ep *channeldb.ChannelEdgePolicy) *edgeLocator {
```
just return non-pointer value.

halseth commented on routing/payment_session.go:
```diff
@@ -35,6 +35,51 @@ type paymentSession struct {
+
+// UpdateEdgePolicy updates edge policy of the routing hints kept in the
+// payment session.
+func (p *paymentSession) UpdateEdgePolicy(el *edgeLocator,
```
also better to take non-pointer for the `edgeLocator` here.

halseth commented on routing/router_test.go:
```diff
@@ -653,6 +653,155 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+
+	const startingBlockHeight = 101
+	ctx, cleanUp, err := createTestCtxFromFile(startingBlockHeight, basicGraphFilePath)
+	defer cleanUp()
```
same comment wrt `cleanup`

halseth commented on routing/router_test.go:
```diff
@@ -653,6 +653,155 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+	// before further processing an edge update. Toggel AssumeChannelValid
+	// to avoid dropping the channel update, as the utxo set query is
+	// not mocked in this test.
+	ctx.router.cfg.AssumeChannelValid = true
```
can be removed

Bluetegu commented:
I did the rebase, but I cannot figure out why the new direction unitest TestSendPaymentPrivateEdgeDirection fails. It didn't fail in the version before the rebase.
I started a new position and unfortunately I no longer have time to work on LND. I tried to fix this issue the last two weekends without success.
I uploaded my latest attempt. The last commit include all the unfinished code.
I hope someone can continue and complete this work (as well as the other of my open PRs), as it contains fixes for two bugs.
I'll do my best to answer questions etc., and can do some work in weekend, but my bandwidth is very limited now.

halseth commented:
@Bluetegu No worries! Thanks for your excellent contributions 👍

halseth commented:
Somebody should pick this up and rebase the still relevant parts.

Roasbeef commented:
Replaced by https://github.com/lightningnetwork/lnd/pull/5332
